docs/source/rfc/isa_restructuring.org

```diff
@@ -0,0 +1,732 @@
+#+TITLE: Iroha Instruction Set Architecture restructuring
+#+AUTHOR: Aleksandr Petrosyan
+#+DATE: 2023-06-25
+
+* Introduction
+
+The term ISA is used in regard to virtual machines and or central
+processing units. Iroha can be regarded as a domain-specific virtual
+machine for blockchain operations and as such the term is applicable
+to the Iroha Special Instructions, which alongside the WASM and the
+expression system comprise an instruction set architecture.
+
+The term reduced instruction set computer (architecture) (RISC) shall
+be used in a purely theoretical capacity, with references to work
+omitted as this is an informal report and recommendation.
+
+The purpose of this document is thus to propose some much needed
+change to the current Complex Instruction Set Computer (architecture)
+based on both a theoretical argument for the scalability of RISC, and
+some practical observations that have been made over the span of
+2021-2023 about the existing and projected uses for the Iroha
+Distributed Ledger Technology. As such it should be treated as a mere
+outline rather than a precise specification.
+
+* Relation to the Executor model
+
+The proposed changes to the Instruction set architecture depend on
+prior art, titled Executor RFC. For some background, Iroha is intended
+as both a ledger for private (permissioned) and public (mistakenly
+known as permissionless) blockchains. The intent was that there would
+be a stack-based virtual machine that would operate on a set of Iroha
+Special Instructions, which shall be shared between the two blockchain
+types, but operate under different semantics, which are chosen at
+compilation time. This was later proven to be insufficient, because
+while almost all public blockchains operate under the same set of
+permissions, private blockchains often not only differ between each
+other, but their permission models also frequently require adjustment
+post-deployment.
+
+Under these circumstances it was decided to make the permission
+validator a dynamic module that is loaded at runtime and is subject to
+upgrade provided that the requisite conditions specified inside the
+module itself are met (e.g. governance vote).  This in turn exposed
+several inefficiencies and some tension between what is considered a
+blockchain invariant and what comprises a blockchain-agnostic,
+universal invariant that must be upheld on all Iroha-based
+blockchains.  To address those inefficiencies it was decided that
+instead of a dynamically loadable permission validator, there would be
+a dynamically loadable executor, which is responsible not only for
+checking if the instruction application violates any of the
+invariants, and checks for permissions, but is responsible wholly for
+upholding the blockchain invariants.
```

Comment:
I'm not very sure about that, becuase it requires users to understand even more things and edge cases given that they already have to know too much to just be able to write validator/executor.

Reply:
That is a problem that is hidden in the old approach also. We just make it explicit.

Reply:
I don't see a problem with this paragraph. Even if blockchain doesn't use executor for upholding blockchain invariants it certainly can do so

Reply:
Addressed it in text.